[Part 1] Add Old Rules for TypeSpecValidation Tool

wanlwanl opened the pull request:
# Problem
- No validation for emitter options in tspconfig.yaml, results in incorrect tspconfig checks in repo and sdk automation failure or generating unexpected result.

# Behavior
- The result of new rules' validations will always be successful, when the actual validation is failed, the TypeSpecValidation tool outputs the actual error messages in stdOutput when single rules suppression is not ready to avoid too much impact on existing services.

wanlwanl committed:
add old rules - part 1

mikeharder commented on eng/tools/typespec-validation/src/tspconfig-rule-helper.ts:
```diff
@@ -0,0 +1,154 @@
+    return this.validate(config);
+  }
+
+  protected skip(_: any): { shouldSkip: boolean; skipResult: RuleResult } {
```
What is the purpose of this method?  It ignores it's parameter and always returns false.

wanlwanl replied:
e.g. some rules to be add will only apply to management plane. It's prepared for these cases

mikeharder commented on eng/tools/typespec-validation/src/tspconfig-rule-helper.ts:
```diff
@@ -0,0 +1,154 @@
+import { join } from "path";
+import { parse as yamlParse } from "yaml";
+
+type ExpectedValueType = string | boolean | RegExp;
```
Are the string and boolean types ever used?  I only see RegExp passed in.

wanlwanl replied:
Prepared to be used for case: `a: true`, `b: "aaa"` - strict equal

mikeharder commented on eng/tools/typespec-validation/src/rules/tspconfig-common-az-service-dir-match-pattern.ts:
```diff
@@ -0,0 +1,12 @@
+      "tspconfig-common-az-service-dir-match-pattern",
+      'Validate "parameters.service-dir.default".',
+      "service-dir",
+      /^sdk\/[^\/]*$/,
```
Is there a reason you used `new RegExp()` above but not here?

wanlwanl replied:
I'm not sure what you are saying, could you rephrase your comment?

mikeharder commented on eng/tools/typespec-validation/src/tspconfig-rule-helper.ts:
```diff
@@ -0,0 +1,154 @@
+    return { shouldSkip: false, skipResult: { success: true } };
+  }
+
+  protected validate(_: any): RuleResult {
```
```suggestion
  protected abstract validate(config: any): RuleResult;
```

mikeharder commented on eng/tools/typespec-validation/src/tspconfig-rule-helper.ts:
```diff
@@ -0,0 +1,154 @@
+
+type ExpectedValueType = string | boolean | RegExp;
+
+class TspconfigRuleBase implements Rule {
```
```suggestion
abstract class TspconfigRuleBase implements Rule {
```

mikeharder commented on eng/tools/typespec-validation/src/tspconfig-rule-helper.ts:
```diff
@@ -0,0 +1,154 @@
+    return { shouldSkip: false, skipResult: { success: true } };
+  }
+
+  protected validate(_: any): RuleResult {
+    return { success: true, stdOutput: `Nothing to validate.` };
+  }
```
```suggestion
  protected abstract validate(config: any): RuleResult;
``````

mikeharder commented on eng/tools/typespec-validation/src/rules/tspconfig-az-java-package-dir-match-pattern.ts:
For organization, please move this file, and the other two `*.ts` files you are adding, to a folder named `rules/tspconfig`.

mikeharder commented on eng/tools/typespec-validation/src/index.ts:
```diff
@@ -1,13 +1,16 @@
 import { FormatRule } from "./rules/format.js";
 import { LinterRulesetRule } from "./rules/linter-ruleset.js";
 import { NpmPrefixRule } from "./rules/npm-prefix.js";
+import { TspConfigCommonAzServiceDirMatchPatternRule } from "./rules/tspconfig-common-az-service-dir-match-pattern.js";
```
How many "tspconfig" rules are you planning to add?  Your "tspconfig" rules are very "narrow", only validating a single property per rule, which makes me concerned the number may grow too large.  The existing rules are more "broad", checking many things in a single "rule".

Looking at the two rules in this PR, having two rules with a shared base class feels over-engineered to me.  I think this could be replaced by a single rule that loads config and checks each of the two properties.

I also think it would be easier for spec authors to understand a single rule like `TspConfigSettings` rather than a large number of rules named `TspConfig*`, some of which may not even apply to their spec?

mikeharder replied:
Later today, I will create another PR showing how I think this could be implemented in a simpler way, and we can compare the two PRs together.

wanlwanl replied:
` also think it would be easier for spec authors to understand a single rule like TspConfigSettings`
So do you mean for main branch, I should add a single rule for all the tspconfig validation I want to add?

wanlwanl replied:
Do you want to seperate one rule for one language or put all of them togethor?
FYI: All the rules I want to add are in this [loop](https://microsoft.sharepoint.com/:fl:/s/f34fd372-ad0b-47ae-9b72-0ee767dfa44e/EVqFBHDGEV1KpRB_SRujBWQBPV0wdukJjcX_K41E9wLSTA?e=0nCIOB&nav=cz0lMkZzaXRlcyUyRmYzNGZkMzcyLWFkMGItNDdhZS05YjcyLTBlZTc2N2RmYTQ0ZSZkPWIlMjFUZWt1bHNPU1BFYVB5NkJaWUYyOXg1QlEyQWxkcndoSG1zWmRDREt4WDlvTHE2S1F4aHR5VDRLekRJODdOWWRJJmY9MDFKWDJJQlBTMlFVQ0hCUlFSTFZGS0tFRDdKRU4yR0JMRSZjPSUyRiZhPUxvb3BBcHAmcD0lNDBmbHVpZHglMkZsb29wLXBhZ2UtY29udGFpbmVyJng9JTdCJTIydyUyMiUzQSUyMlQwUlRVSHh0YVdOeWIzTnZablF1YzJoaGNtVndiMmx1ZEM1amIyMThZaUZVWld0MWJITlBVMUJGWVZCNU5rSmFXVVl5T1hnMVFsRXlRV3hrY25kb1NHMXpXbVJEUkV0NFdEbHZUSEUyUzFGNGFIUjVWRFJMZWtSSk9EZE9XV1JKZkRBeFNsZ3lTVUpRVWxGRFVqZE5Ua2N5V1UxS1JsbEZNMDR6VWxaSVNVOUJOMWslM0QlMjIlMkMlMjJpJTIyJTNBJTIyMzA0MDI2MDktODdlYy00NDdhLWFiOGQtNDUzNDRkODNjZDY2JTIyJTdE). Could you review it and share your idea?

mikeharder replied:
Thanks, the additional context helps.  I will review that doc to understand the full set of rules you are planning and come up with a recommendation that can work in both the current implementation and support ESLint should we move there.  I'll get back to you by Monday.

wanlwanl replied:
@mikeharder Please review this PR, already updated.

sample output:
```
Executing rule: SdkTspConfigValidation
[SdkTspConfigValidation]: validation failed.
- The value of options.@azure-tools/typespec-csharp.namespace "Azure.AI.DocumentIntelligence" does not match "{package-dir}". Please update the value of "options.@azure-tools/typespec-csharp.namespace" to match "{package-dir}".
- Failed to find "options.@azure-tools/typespec-csharp.clear-output-folder". Please add "options.@azure-tools/typespec-csharp.clear-output-folder".
```

wanlwanl committed:
Update eng/tools/typespec-validation/src/tspconfig-rule-helper.ts

add abstract

Co-authored-by: Mike Harder <mharder@microsoft.com>

wanlwanl committed:
Update eng/tools/typespec-validation/src/tspconfig-rule-helper.ts

Co-authored-by: Mike Harder <mharder@microsoft.com>

wanlwanl committed:
Merge remote-tracking branch 'Azure/main' into wanl/add-old-rule

wanlwanl committed:
WIP

wanlwanl committed:
put all tspconfig validation in a single rule

sample output:
Executing rule: SdkTspConfigValidation
[SdkTspConfigValidation]: validation failed.
- The value of options.@azure-tools/typespec-csharp.namespace "Azure.AI.DocumentIntelligence" does not match "{package-dir}". Please update the value of "options.@azure-tools/typespec-csharp.namespace" to match "{package-dir}".
- Failed to find "options.@azure-tools/typespec-csharp.clear-output-folder". Please add "options.@azure-tools/typespec-csharp.clear-output-folder".

wanlwanl committed:
rename

mikeharder commented on eng/tools/typespec-validation/src/tspconfig-rule-helper.ts:
I would probably move this code into file `sdk-tspconfig-validation.ts` -- I don't see any reason to split it into two files.

mikeharder committed:
Merge branch 'main' into wanl/add-old-rule
